src/fstext/fstext-utils.h

```diff
@@ -392,7 +392,23 @@ void RhoCompose(const Fst<Arc> &fst1,
                 typename Arc::Label rho_label,
                 MutableFst<Arc> *fst);
 
-
+// Lookahead composition is used for optimized online
+// composition of FSTs during decoding. See
+// nnet3/nnet3-latgen-faster-lookahead.cc
+template<class Arc>
+ComposeFst<Arc> LookaheadComposeFst(const Fst<Arc> &ifst1,
+                                    const Fst<Arc> &ifst2) {
+  typedef LookAheadMatcher< StdFst > M;
+  typedef AltSequenceComposeFilter<M> SF;
+  typedef LookAheadComposeFilter<SF, M>  LF;
+  typedef PushWeightsComposeFilter<LF, M> WF;
+  typedef PushLabelsComposeFilter<WF, M> ComposeFilter;
+
+  fst::CacheOptions cache_opts(true, 1 << 26LL);
+  ComposeFstOptions<StdArc, M, ComposeFilter> opts(cache_opts);
```

Comment:
Hi @nshmyrev this is very interesting. I read a paper called [A Generalized Composition Algorithm for Weighted Finite-State Transducers](https://static.googleusercontent.com/media/research.google.com/en//pubs/archive/35539.pdf) about a month ago. It mentions that its dynamic runtime composition is provided in the OpenFST library. Does `ComposeFst` implement its functionality?

In the introduction, you can read about 3 problems the paper discusses with "normal" composition of HCL and G on the fly in section 1.

Reply:
Yes, this is the implementation of the described paper from openfst integrated with kaldi decoding.